monkey/agent_plugins/exploiters/mssql/src/plugin.py

```diff
@@ -103,15 +93,50 @@ def run(
                 tcp_port_selector=self._tcp_port_selector,
             )
 
-            mssql_client = StubMSSQLClient()
+            mssql_client = MSSQLClient(mssql_options.server_timeout)
+
+            mssql_exploit_client = MSSQLExploitClient(
+                exploiter_name=self._plugin_name,
+                agent_id=self._agent_id,
+                agent_event_publisher=self._agent_event_publisher,
+                mssql_client=mssql_client,
+            )
 
             mssql_exploiter = MSSQLExploiter(
-                self._agent_id, mssql_client, agent_binary_server_factory, self._otp_provider
+                self._agent_id,
+                mssql_exploit_client,
+                agent_binary_server_factory,
+                self._otp_provider,
             )
             return mssql_exploiter.exploit_host(
-                host, servers, current_depth, mssql_options, brute_force_credentials, interrupt
+                host,
+                mssql_options,
+                servers,
+                current_depth,
+                brute_force_credentials,
+                ports_to_try,
+                interrupt,
             )
         except Exception as err:
             msg = f"An unexpected exception occurred while attempting to exploit host: {err}"
             logger.exception(msg)
             return ExploiterResultData(error_message=msg)
+
+
+def get_ports_to_try(host: TargetHost, options: MSSQLOptions) -> Set[NetworkPort]:
+    tcp_ports = host.ports_status.tcp_ports
+    ports_to_try = {port for port in set(options.target_ports) if port not in tcp_ports.closed}
+
+    if options.try_discovered_mssql_ports:
+        mssql_tcp_ports = {
+            port for port in tcp_ports.open if tcp_ports[port].service == NetworkService.MSSQL
+        }
+        ports_to_try.update(mssql_tcp_ports)
+
+    if options.try_discovered_unknown_ports:
+        unknown_tcp_ports = {
+            port for port in tcp_ports.open if tcp_ports[port].service == NetworkService.UNKNOWN
+        }
+        ports_to_try.update(unknown_tcp_ports)
```

Comment:
NIT:

Alternatively:
```suggestion
    services_to_scan = {}
    if options.try_discovered_mssql_ports:
        services_to_scan.add(NetworkService.MSSQL)
        
    if options.try_discovered_unknown_ports:
        services_to_scan.add(NetworkService.UNKNOWN)
    
    additional_ports = {
        p for p in tcp_ports.open if tcp_ports[p].service in services_to_scan
    }
    ports_to_try.update(additional_ports)
```